cmd/containerd-shim-runc-v2/task/service.go

```diff
@@ -707,18 +682,81 @@ func (s *service) send(evt interface{}) {
 	s.events <- evt
 }
 
-// s.mu must be locked when calling handleProcessExit
-func (s *service) handleProcessExit(e runcC.Exit, c *runc.Container, p process.Process) {
-	if ip, ok := p.(*process.Init); ok {
-		// Ensure all children are killed
-		if runc.ShouldKillAllOnExit(s.context, c.Bundle) {
-			if err := ip.KillAll(s.context); err != nil {
-				log.G(s.context).WithError(err).WithField("id", ip.ID()).
-					Error("failed to kill init's children")
+// handleInitExit processes container init process exits.
+// This is handled separately from non-init exits, because there
+// are some extra invariants we want to ensure in this case, namely:
+// - for a given container, the init process exit MUST be the last exit published
+// This is achieved by:
+// - waiting for all pending execs to either successfully start or early-exit
+// (this happens before handleInitExit is called)
+// - killing all running container processes
+// - removing all processes from s.running so that we s.processExits doesn't
+// publish their exits
+// - subscribing to exits+waiting for the running process count (for a
+// container) to go to 0, or timing out.
+// - publishing the received exec exits, and then publishing the init exit.
+func (s *service) handleInitExit(e runcC.Exit, c *runc.Container, p *process.Init) {
+	// kill all running container processes
+	if err := p.KillAll(s.context); err != nil {
+		log.G(s.context).WithError(err).WithField("id", p.ID()).
+			Error("failed to kill init's children")
+	}
+
+	go func() {
+		timeout := 20 * time.Millisecond
+		timer := time.NewTimer(timeout)
+		var numRunningExecs int
+		defer timer.Stop()
+		// wait for running processes to exit
```

Comment:
We can do better than periodic polling. How about pub-sub of events whenever `s.runningExecs[c]` gets decremented?
```go
type service struct {
	// ...
	execCountSubscribers map[*runc.Container]chan<- int
	// ...
}
```
```go
func (s *service) handleProcessExit(e runcC.Exit, c *runc.Container, p process.Process) {
	// ...
	if _, init := p.(*process.Init); !init {
		s.lifecycleMu.Lock()
		s.runningExecs[c]--
		if ch, ok := s.execCountSubscribers[c]; ok {
			ch <- s.runningExecs[c]
		}
		s.lifecycleMu.Unlock()
	}
}
```
```go
func (s *service) handleInitExit(e runcC.Exit, c *runc.Container, p *process.Init) {
	// ...
	s.lifecycleMu.Lock()
	defer s.lifecycleMu.Unlock()
	numRunningExecs := s.runningExecs[c]
	if numRunningExecs == 0 {
		return
	}
	events := make(chan int, numRunningExecs)
	s.execCountSubscribers[c] = events

	go func() {
		// ...
		defer func() {
			s.lifecycleMu.Lock()
			defer s.lifecycleMu.Unlock()
			delete(s.execCountSubscribers, c)
		}()
WAIT:
		for {
			select {
			case <-watchdog.C:
				// handle watchdog timeout
			case numRunningExecs = <-events:
				if numRunningExecs == 0 {
					break WAIT
				}
				// kick watchdog timer
			}
		}
		// ...
	}()
}
```

Alternatively, a `chan struct{}` which is closed when `s.runningExecs[c]` drops to zero would also work, though without the watchdog kicking.

Reply:
Oh, yeah! I really like this, and it lessens my concerns about lock contention for `s.lifecycleMu`!